src/locator.rs

```diff
@@ -65,55 +63,84 @@ macro_rules! locator {
     };
 }
 
-/// Core, and most services that interact with Core,
-/// refer to open source packages via the `Locator` type.
+/// The regular expression used to parse the locator.
 ///
-/// This type is nearly universally rendered to a string
-/// before being serialized to the database or sent over the network.
+/// ```
+/// # use locator::locator_regex;
 ///
-/// This type represents a _validly-constructed_ `Locator`, but does not
-/// validate whether a `Locator` is actually valid. This means that a
-/// given `Locator` is guaranteed to be correctly formatted data,
-/// but that the actual repository or revision to which the `Locator`
-/// refers is _not_ guaranteed to exist or be accessible.
-/// Currently the canonical method for validating whether a given `Locator` is
-/// accessible is to run it through the Core fetcher system.
+/// // Get the raw string used for the expression.
+/// let expression = locator_regex!();
+///
+/// // Parse the regular expression.
+/// // The expression is compiled once per callsite.
+/// let parsed = locator_regex!(parse => "git+github.com/fossas/locator-rs$v2.2.0");
+/// ```
+#[macro_export]
+#[doc(hidden)]
+macro_rules! locator_regex {
+    () => {
+        r"^(?:([a-z-]+)\+|)([^$]+)(?:\$|)(.+|)$"
+    };
+    (parse => $input:expr) => {
+        lazy_regex::regex_captures!(r"^(?:([a-z-]+)\+|)([^$]+)(?:\$|)(.+|)$", $input)
+    };
+}
```

Comment:
[Question] Is the reasoning for using a macro here instead of a function that you also want the result of `regex_captures!` to be inlined rather than have to call out to an auxiliary function?

To me it feels simpler to just have a `const` and a function-like thing (even a macro with no keyword) for parsing rather than having to learn a new syntax for invoking this.

Reply:
Effectively: the point of macros is:
- if you want something done at compile time (like this, or `const_assert!`, etc)
- if you want to introduce flexibility to call sites (like `format!()`)

Inlining isn't actually a reason to use a macro: the Rust compiler will automatically inline functions that are able to be inlined (and can be nudged in that direction with the `#[inline]` directive as well, although it's generally not needed)

Reply:
I think I understand the reasoning behind `regex_captures` being like this, but my question is more about `locator_regex`. It seems like something like:

```rust
const LOCATOR_REGEX: &str = r"^(?:([a-z-]+)\+|)([^$]+)(?:\$|)(.+|)$";

/// parse locator
pub fn parse_locator_regex(input: &str) -> Option<(&str, &str, &str, &str)> {
    lazy_regex::regex_captures!(r"^(?:([a-z-]+)\+|)([^$]+)(?:\$|)(.+|)$", input)
}
```
would be equivalent (aside from `parse_locator_regex` being a function call) and more straightforward to understand.  Or even two separate macros would work.

To me, it seems like the current `locator_regex` can behave like a constant function (or value) if you use it with no arguments but can _also_ behave almost like a function with one argument - except you have to use a special calling syntax: `locator_regex!(parse => "input text");`. I guess it's fairly obvious what it does based on the `parse =>` symbols, but I don't see much gain from adopting it. To me, it seems that the disambiguation that the macro rule does with `parse =>` is logically the same as having a distinct name for the second case of the macro.

There's also the possibility that this is much more idiomatic in Rust than I understand yet and I just need to get more comfortable with it.

Also, I know I've pressed on this a lot, if it ever feels like I'm just beating a dead horse please let me know because that's absolutely not what I'm trying to do.

Reply:
> I think I understand the reasoning behind `regex_captures` being like this, but my question is more about `locator_regex`. It seems like something like:
> 
> ```rust
> const LOCATOR_REGEX: &str = r"^(?:([a-z-]+)\+|)([^$]+)(?:\$|)(.+|)$";
> 
> /// parse locator
> pub fn parse_locator_regex(input: &str) -> Option<(&str, &str, &str, &str)> {
>     lazy_regex::regex_captures!(r"^(?:([a-z-]+)\+|)([^$]+)(?:\$|)(.+|)$", input)
> }
> ```
> 
> would be equivalent (aside from `parse_locator_regex` being a function call) and more straightforward to understand. Or even two separate macros would work.

Yes, this is equivalent.

> To me, it seems like the current `locator_regex` can behave like a constant function (or value) if you use it with no arguments but can _also_ behave almost like a function with one argument - except you have to use a special calling syntax: `locator_regex!(parse => "input text");`. I guess it's fairly obvious what it does based on the `parse =>` symbols, but I don't see much gain from adopting it. To me, it seems that the disambiguation that the macro rule does with `parse =>` is logically the same as having a distinct name for the second case of the macro.

To me, the primary benefit is:
1. Having them in the same macro _forces_ colocation.
2. Having them in the same macro makes it clear that these would ideally share the same string literal, they just can't do so right now- making it a prime spot for us to figure out a way to better handle this in the future. Also, any such future functionality would by necessity be a macro itself, so this at least has a decent chance to minimize refactors later when/if we make this better.

> There's also the possibility that this is much more idiomatic in Rust than I understand yet and I just need to get more comfortable with it.

I think it's less that either direction is idiomatic and more that either direction is equally valid based on the context. I think the main difference between our viewpoints _seems_ to be that I don't consider macros to be "less ideal, but use them if you have to"- I consider them to be a neutral tool in the toolbox; it seems like you consider them to be less ideal. I'm definitely open to being convinced that my approach to macros is wrong but they're just so heavily used in Rust (since there's no reflection) that trying to label macros second class citizens feels like a losing battle to me.

> Also, I know I've pressed on this a lot, if it ever feels like I'm just beating a dead horse please let me know because that's absolutely not what I'm trying to do.

Never a problem! If we haven't truly reached consensus here I'm glad you're bringing it up. I'd rather we're all on the same page, whatever that page is.

Reply:
Hey, I just wanted to let you know I plan to respond to this but just haven't gotten the chance yet - I appreciate you taking the time write this and am planning to come back to it.